File: java/src/main/java/com/murex/tbw/finance/CurrencyConverter.java
/*******************************************************************************
 *
 * Copyright (c) {2003-2019} Murex S.A.S. and its affiliates.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the MIT License
 * which accompanies this distribution, and is available at
 * https://opensource.org/licenses/MIT
 *
 *******************************************************************************/

package com.murex.tbw.finance;

import com.google.common.collect.ImmutableMap;
import com.murex.tbw.domain.country.Currency;

import java.util.Map;

import static com.murex.tbw.domain.country.Currency.*;

public final class CurrencyConverter {

    private static final Map<Currency, Double> EXCHANGE_RATE = ImmutableMap.<Currency, Double>builder()
            .put(US_DOLLAR, 1.0d)
            .put(EURO, 1.14d)
            .put(POUND_STERLING, 1.27d)
            .put(RENMINBI, 0.15d)
            .put(YEN, 0.0093d)
            .put(AUSTRALIAN_DOLLAR, 0.7d)
            .build();

    public static double fromUSD(double input, Currency currency) {
        return input / EXCHANGE_RATE.get(currency);
    }

    public static double toUSD(double input, Currency currency) {
        return input * EXCHANGE_RATE.get(currency);
    }
}

File: typescript/src/com/murex/tbw/finance/CurrencyConverter.ts
import * as Immutable from "immutable";
import { Currency } from "../domain/country/Currency";

const EXCHANGE_RATES = Immutable.Map<Currency, number>(
  new Map<Currency, number>([
    [Currency.Euro, 1.14],
    [Currency.Dollar, 1.0],
    [Currency.PoundSterling, 1.27],
    [Currency.Renminbi, 0.15],
    [Currency.Yen, 0.0093],
    [Currency.AustralianDollar, 0.7],
  ])
);

export function fromUSD(input: number, currency: Currency): number {
  const exchangeRate = getExchangeRateOrThrow(currency);

  return input / exchangeRate;
}

export function toUSD(input: number, currency: Currency): number {
  const exchangeRate = getExchangeRateOrThrow(currency);

  return input * exchangeRate;
}

function getExchangeRateOrThrow(currency: Currency) {
  const exchangeRate = EXCHANGE_RATES.get(currency);
  if (!exchangeRate)
    throw new Error("Unexpected currency " + currency.toString());

  return exchangeRate || 0;
}
